src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexShard.java

```diff
@@ -0,0 +1,124 @@
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.index;
+
+import com.amazon.opendistroforelasticsearch.knn.index.codec.KNNCodecUtil;
+import com.amazon.opendistroforelasticsearch.knn.index.v1736.KNNIndex;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.FilterLeafReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.index.SegmentReader;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardPath;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * KNNIndexShard wraps IndexShard and adds methods to perform k-NN related operations against the shard
+ */
+public class KNNIndexShard {
+    private IndexShard indexShard;
+    private KNNIndexCache knnIndexCache;
+
+    private static Logger logger = LogManager.getLogger(KNNIndexShard.class);
+
+    /**
+     * Constructor to generate KNNIndexShard. We do not perform validation that the Index the IndexShard is from
+     * is in fact a k-NN Index (index.knn = true). This may make sense to add later, but for now the operations for
+     * KNNIndexShards that are not from a k-NN index should be no-ops.
+     *
+     * @param indexShard IndexShard to be wrapped.
+     */
+    public KNNIndexShard(IndexShard indexShard) {
+        this.indexShard = indexShard;
+        this.knnIndexCache = KNNIndexCache.getInstance();
+    }
+
+    /**
+     * Return the underlying IndexShard
+     *
+     * @return IndexShard
+     */
+    public IndexShard getIndexShard() {
+        return indexShard;
+    }
+
+    /**
+     * Return the name of the shards index
+     *
+     * @return Name of shard's index
+     */
+    public String getIndexName() {
+        return indexShard.shardId().getIndexName();
+    }
+
+    /**
+     * Load all of the HNSW graphs for this shard into the cache. Note that getIndices is called to prevent loading
+     * in duplicates.
+     *
+     * @return a List of KNNIndex's from this shard that are in the cache after this operation.
+     * @throws IOException Thrown when getting the HNSW Paths to be loaded in
+     */
+    public List<KNNIndex> warmup() throws IOException {
+        logger.info("[KNN] Warming up index: " + getIndexName());
+        Engine.Searcher searcher = indexShard.acquireSearcher("knn-warmup");
+        List<KNNIndex> indices;
+        try {
+            indices = knnIndexCache.getIndices(getHNSWPaths(searcher.getIndexReader()), getIndexName());
+            searcher.close();
```

Comment:
move ```searcher.close()``` to finally block and you could remove from ```catch```

Reply:
Good suggestion. Will update